examples/structured_data/collaborative_filtering_movielens.py

```diff
@@ -0,0 +1,212 @@
+        super(RecommenderNet, self).__init__(**kwargs)
+        self.num_users = num_users
+        self.num_movies = num_movies
+        self.embedding_size = embedding_size
+        self.user_embedding = layers.Embedding(
+            num_users,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=keras.regularizers.l2(1e-6),
+        )
+        self.user_bias = layers.Embedding(num_users, 1)
+        self.movie_embedding = layers.Embedding(
+            num_movies,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=keras.regularizers.l2(1e-6),
+        )
+        self.movie_bias = layers.Embedding(num_movies, 1)
+
+    def call(self, inputs):
+        user_vector = self.user_embedding(inputs[:, 0])
+        user_bias = self.user_bias(inputs[:, 0])
+        movie_vector = self.movie_embedding(inputs[:, 1])
+        movie_bias = self.movie_bias(inputs[:, 1])
+        dot_user_movie = tf.tensordot(user_vector, movie_vector, 2)
+        # Add all the components (including bias)
+        x = dot_user_movie + user_bias + movie_bias
+        # The sigmoid activation forces the rating to between 0 and 1
+        return tf.nn.sigmoid(x)
+
+
+inputs = layers.Input(shape=(2,))
+model = RecommenderNet(num_users, num_movies, EMBEDDING_SIZE)
+outputs = model(inputs)
+opt = keras.optimizers.Adam(lr=0.001)
+model.compile(loss=tf.keras.losses.BinaryCrossentropy(), optimizer=opt)
+
+"""
+## Train the model based on the data split
+"""
+history = model.fit(
+    x=x_train,
+    y=y_train,
+    batch_size=64,
+    epochs=5,
+    verbose=1,
+    validation_data=(x_test, y_test),
+)
+
+"""
+## Plot training and test loss
+"""
+plt.plot(history.history["loss"])
+plt.plot(history.history["val_loss"])
+plt.title("model loss")
+plt.ylabel("loss")
+plt.xlabel("epoch")
+plt.legend(["train", "test"], loc="upper left")
+plt.show()
+
+"""
+## Show top 10 movie recommendations to an user
+"""
+
+movie_df = pd.read_csv(movielens_dir / "movies.csv")
+
+# Let us get an user and see the top recommendations.
+user_id = df.userId.sample(1).iloc[0]
+movies_watched_by_user = df[df.userId == user_id].movieId.values
+movies_not_watched = movie_df[~movie_df["movieId"].isin(movies_watched_by_user)][
+    "movieId"
+]
+movies_not_watched = list(
+    set(movies_not_watched).intersection(set(movie2movie_encoded.keys()))
+)
+movies_not_watched = [[movie2movie_encoded.get(x)] for x in movies_not_watched]
+user_encoder = user2user_encoded.get(user_id)
+user_movie_array = np.hstack(
+    ([[user_id]] * len(movies_not_watched), movies_not_watched)
+)
+ratings = model.predict(user_movie_array).flatten()
+top_ratings_indices = ratings.argsort()[-10:][::-1]
+recommended_movie_ids = [movie_encoded2movie.get(x) for x in top_ratings_indices]
+
+print("Here are the top 10 movie recommendations for user: {}".format(user_id))
```

Comment:
Please also print the titles of the movies were rated highly by the user, so the readers can see the relationship between liked movies and the new recommendations

Reply:
Adding title + movie genre too for better understanding. Had a bug in the code when looking up movie ids that is resolved